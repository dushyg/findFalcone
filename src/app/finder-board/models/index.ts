export { IFalconAppState } from './falconApp.state';
export { RawPlanet, IPlanet } from './planet';
export { RawVehicle, IVehicle } from './vehicle';
export { default as PlanetChange } from './planetChange';
export { default as VehicleChange } from './vehicleChange';
export { ISearchAttempt } from './searchAttempt';
