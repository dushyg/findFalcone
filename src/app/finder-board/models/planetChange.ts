import { IPlanet } from './planet';

export default class PlanetChange {
  constructor(public widgetId: number, public newPlanetName: string) {}
}
