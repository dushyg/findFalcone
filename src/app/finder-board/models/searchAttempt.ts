import { IVehicle } from "./vehicle";
import { IPlanet } from "./planet";

export interface ISearchAttempt {
  searchedPlanet: string;
  vehicleUsed: string;
}
