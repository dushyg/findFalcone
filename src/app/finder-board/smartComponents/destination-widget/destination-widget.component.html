<div
  class="destinationWidget"
  [attr.widgetId]="widgetId"
  *ngIf="planetListInitialized"
>
  <span>Destination {{ widgetId }}</span>
  <app-typeahead
    [sourceArray]="planetList"
    [resetTypeAhead$]="resetTypeAhead$"
    (itemSelected)="planetSelected($event)"
    [widgetId]="widgetId"
  >
  </app-typeahead>

  <app-vehicle-list
    *ngIf="lastSelectedPlanet != 'Select'"
    [destinationDistance]="destinationDistance"
    [widgetId]="widgetId"
  >
  </app-vehicle-list>
</div>
