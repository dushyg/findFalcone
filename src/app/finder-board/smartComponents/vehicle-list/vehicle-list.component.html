<div class="vehicleList">
  <div *ngFor="let vehicle of vehicleList; let i = index">
    <input
      type="radio"
      name="radio_{{ widgetId }}"
      [value]="vehicle.name"
      [ngModel]="lastSelectedVehicle.name"
      [disabled]="
        (vehicle && vehicle.maxDistance < destinationDistance) ||
        vehicle.availNumUnits == 0
      "
      id="vehicle_{{ widgetId }}_{{ i }}"
      (change)="vehicleSelected(vehicle)"
    />&nbsp;
    <label for="vehicle_{{ widgetId }}_{{ i }}">
      {{ vehicle.name + " (" + vehicle.availNumUnits + ")" }}
    </label>

    <br />
  </div>
</div>
