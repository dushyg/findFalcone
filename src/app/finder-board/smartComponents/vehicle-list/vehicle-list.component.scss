.vehicleList {
  display: inline-block;
  text-align: left;
}

input[type="radio"]:disabled {
  cursor: not-allowed;
}

input[type="radio"]:disabled + label {
  cursor: not-allowed;
  text-decoration: line-through;
  color: lightcoral;
}

input[type="radio"]:checked + label {
  font-weight: bold;
  font-style: italic;
  text-decoration: none;
  color: black;
}
