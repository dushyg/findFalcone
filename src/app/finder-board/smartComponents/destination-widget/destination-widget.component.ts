import { Component, OnInit, OnDestroy } from "@angular/core";

import { IPlanet } from "src/app/finder-board/models/planet";
import PlanetChange from "src/app/finder-board/models/planetChange";
import { Observable, Subject, combineLatest } from "rxjs";
import { FinderFacadeService } from "src/app/finder-board/services/finder-facade.service";
import { takeWhile, withLatestFrom } from "rxjs/operators";

@Component({
  selector: "app-destination-widget",
  templateUrl: "./destination-widget.component.html",
  styleUrls: ["./destination-widget.component.scss"],
})
export class DestinationWidgetComponent implements OnInit, OnDestroy {
  private static createdWidgetCount: number = 0;
  public widgetId: number;

  public planetList: IPlanet[];

  public destinationDistance: number = 0;

  public lastSelectedPlanet: string = "Select";

  private resetTypeAheadSubject = new Subject<void>();
  public resetTypeAhead$: Observable<
    void
  > = this.resetTypeAheadSubject.asObservable();

  private isComponentActive = true;
  public planetListInitialized = false;
  constructor(private finderFacadeService: FinderFacadeService) {
    // using modulo operator to cycle the widget ids from 1 to max widget count
    this.widgetId =
      (DestinationWidgetComponent.createdWidgetCount %
        this.finderFacadeService.getCountOfWidgetsDisplayed()) +
      1;
    DestinationWidgetComponent.createdWidgetCount++;
  }

  ngOnInit(): void {
    combineLatest(
      this.finderFacadeService.unsearchedPlanets$,
      this.finderFacadeService.isLoading$
    )
      .pipe(takeWhile(() => !this.planetListInitialized))
      .subscribe(([unsearchedPlanets, isLoading]) => {
        if (unsearchedPlanets && !isLoading) {
          this.planetList = unsearchedPlanets;
          this.planetListInitialized = true;
        }
      });

    this.finderFacadeService.planetChangedAction$
      .pipe(
        withLatestFrom(this.finderFacadeService.unsearchedPlanets$),
        takeWhile(() => this.isComponentActive)
      )
      .subscribe(([planetChange, unsearchedPlanets]) => {
        if (planetChange && planetChange.widgetId < this.widgetId) {
          this.planetList = unsearchedPlanets;
          this.clearLastSelection();
        }
      });

    this.finderFacadeService.vehicleChangedAction$
      .pipe(
        withLatestFrom(this.finderFacadeService.unsearchedPlanets$),
        takeWhile(() => this.isComponentActive)
      )
      .subscribe(([vehicleChange, unsearchedPlanets]) => {
        if (vehicleChange && vehicleChange.widgetId < this.widgetId) {
          this.planetList = unsearchedPlanets;
          this.clearLastSelection();
        }
      });
  }

  clearLastSelection(): void {
    this.lastSelectedPlanet = "Select";
    this.resetTypeAheadSubject.next();
  }

  public planetSelected(selectedPlanet: IPlanet) {
    this.destinationDistance = selectedPlanet.distance;

    this.lastSelectedPlanet = selectedPlanet.name;

    this.finderFacadeService.planetChanged(
      new PlanetChange(this.widgetId, selectedPlanet.name)
    );
  }

  ngOnDestroy(): void {
    this.isComponentActive = false;
  }
}
