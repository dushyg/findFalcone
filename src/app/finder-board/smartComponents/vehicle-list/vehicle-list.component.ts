import {
  Component,
  OnInit,
  Input,
  OnDestroy,
  ChangeDetectionStrategy,
  ChangeDetectorRef,
} from '@angular/core';
import { IVehicle } from 'src/app/finder-board/models/vehicle';
import VehicleChange from 'src/app/finder-board/models/vehicleChange';
import { FinderFacadeService } from 'src/app/finder-board/services/finder-facade.service';
import { takeWhile, withLatestFrom } from 'rxjs/operators';

@Component({
  selector: 'app-vehicle-list',
  templateUrl: './vehicle-list.component.html',
  styleUrls: ['./vehicle-list.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class VehicleListComponent implements OnInit, OnDestroy {
  private isComponentActive = true;

  constructor(
    private finderFacadeService: FinderFacadeService,
    private changeDetector: ChangeDetectorRef
  ) {}

  public vehicleList: IVehicle[];

  @Input() public destinationDistance: number;
  @Input() public widgetId: number;

  public lastSelectedVehicle: IVehicle = {
    availNumUnits: 0,
    name: null,
  } as IVehicle;

  ngOnInit() {
    let vehicleListInitialized = false;

    this.finderFacadeService.vehicleInventory$
      .pipe(takeWhile(() => !vehicleListInitialized))
      .subscribe((initialVehicles: IVehicle[]) => {
        if (initialVehicles) {
          this.vehicleList = initialVehicles;
          vehicleListInitialized = true;
          this.changeDetector.detectChanges();
        }
      });

    this.finderFacadeService.vehicleChangedAction$
      .pipe(
        withLatestFrom(this.finderFacadeService.vehicleInventory$),
        takeWhile(() => this.isComponentActive)
      )
      .subscribe(([vehicleChange, updatedVehicles]) => {
        if (vehicleChange && vehicleChange.widgetId <= this.widgetId) {
          this.vehicleList = updatedVehicles;
          this.changeDetector.detectChanges();
        }
      });

    this.finderFacadeService.planetChangedAction$
      .pipe(
        withLatestFrom(this.finderFacadeService.vehicleInventory$),
        takeWhile(() => this.isComponentActive)
      )
      .subscribe(([planetChange, updatedVehicles]) => {
        // need to refresh the vehicle list even when the planet is chnaged for current widget
        if (planetChange && planetChange.widgetId <= this.widgetId) {
          this.vehicleList = updatedVehicles;
          if (planetChange.widgetId !== this.widgetId) {
            this.clearLastSelection();
            this.changeDetector.detectChanges();
          }
        }
      });
  }

  public vehicleSelected(vehicle: IVehicle) {
    this.lastSelectedVehicle = vehicle;

    this.finderFacadeService.vehicleChanged(
      new VehicleChange(this.widgetId, vehicle.name)
    );
  }

  clearLastSelection(): void {
    this.lastSelectedVehicle = { availNumUnits: 0, name: null } as IVehicle;
  }

  ngOnDestroy(): void {
    this.isComponentActive = false;
  }
  // tslint:disable-next-line: use-lifecycle-interface
  ngAfterViewChecked(): void {
    // console.log('VehicleListComponent ngAfterViewChecked', this.widgetId);
  }
}
