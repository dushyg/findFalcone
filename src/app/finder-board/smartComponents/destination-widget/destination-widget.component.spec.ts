import { ComponentFixture, TestBed } from '@angular/core/testing';
import { DestinationWidgetComponent } from './destination-widget.component';
import { TypeaheadComponent } from '../../presenterComponents/typeahead/typeahead.component';
import { VehicleListComponent } from '../vehicle-list/vehicle-list.component';
import { createSpyObj } from '../../../testingUtility';
import { FinderFacadeService } from '../../services/finder-facade.service';
import { PlanetsService } from '../../services/planets.service';
import { VehiclesService } from '../../services/vehicles.service';
import { FalconeTokenService } from '../../services/falcone-token.service';
import { Router } from '@angular/router';
import { NO_ERRORS_SCHEMA, Type, ChangeDetectorRef } from '@angular/core';
import { By } from '@angular/platform-browser';
import { of } from 'rxjs';
import { IPlanet, IVehicle } from '../../models';

describe('DestinationWidgetComponent', () => {
  let component: DestinationWidgetComponent;
  let fixture: ComponentFixture<DestinationWidgetComponent>;

  let planetServiceMock: { [key: string]: jest.Mock<any, any> };
  let vehiclesServiceMock: { [key: string]: jest.Mock<any, any> };
  let apiTokenServiceMock: { [key: string]: jest.Mock<any, any> };
  let routerServiceMock: { [key: string]: jest.Mock<any, any> };
  let planetListToBeReturned: IPlanet[];
  let vehicleListToBeReturned: IVehicle[];
  const apiTokenToBeReturned = { token: 'plmVHX' };
  beforeEach(() => {
    planetServiceMock = createSpyObj(['getAllPlanets']);
    vehiclesServiceMock = createSpyObj(['getAllVehicles']);
    apiTokenServiceMock = createSpyObj(['getFalconeFinderApiToken']);
    routerServiceMock = createSpyObj(['navigate']);

    TestBed.configureTestingModule({
      declarations: [
        DestinationWidgetComponent,
        TypeaheadComponent,
        VehicleListComponent,
      ],
      providers: [
        FinderFacadeService,
        { provide: PlanetsService, useValue: planetServiceMock },
        { provide: VehiclesService, useValue: vehiclesServiceMock },
        { provide: FalconeTokenService, useValue: apiTokenServiceMock },
        { provide: Router, useValue: routerServiceMock },
      ],
      schemas: [NO_ERRORS_SCHEMA],
    });

    planetListToBeReturned = [
      { name: 'Donlon', distance: 100 },
      { name: 'Enchai', distance: 200 },
      { name: 'Jebing', distance: 300 },
      { name: 'Sapir', distance: 400 },
    ];

    vehicleListToBeReturned = [
      {
        name: 'Space pod',
        availNumUnits: 2,
        maxDistance: 200,
        speed: 2,
        totalNumUnits: 2,
      },
      {
        name: 'Space rocket',
        availNumUnits: 1,
        maxDistance: 300,
        speed: 4,
        totalNumUnits: 1,
      },
      {
        name: 'Space shuttle',
        availNumUnits: 1,
        maxDistance: 400,
        speed: 5,
        totalNumUnits: 1,
      },
      {
        name: 'Space ship',
        availNumUnits: 2,
        maxDistance: 600,
        speed: 10,
        totalNumUnits: 2,
      },
    ];

    planetServiceMock.getAllPlanets.mockReturnValue(of(planetListToBeReturned));
    vehiclesServiceMock.getAllVehicles.mockReturnValue(
      of(vehicleListToBeReturned)
    );
    apiTokenServiceMock.getFalconeFinderApiToken.mockReturnValue(
      of(apiTokenToBeReturned)
    );

    fixture = TestBed.createComponent(DestinationWidgetComponent);

    component = fixture.componentInstance;
  });

  it('should create', () => {
    fixture.detectChanges();

    expect(component).toBeTruthy();
  });

  it('should render typeahead component', () => {
    fixture.detectChanges();
    const facadeService = fixture.componentRef.injector.get(
      FinderFacadeService
    );

    facadeService.initializeAppData();
    fixture.detectChanges();

    const typeAheadDebugElement = fixture.debugElement.query(
      By.directive(TypeaheadComponent)
    );
    expect(typeAheadDebugElement).toBeTruthy();

    const typeAheadComponent = typeAheadDebugElement.componentInstance as TypeaheadComponent;
    expect(typeAheadComponent.filteredSourceArray).toEqual(
      typeAheadComponent.sourceArray
    );

    expect(typeAheadComponent.doShowResults).toBeFalsy();
    expect(typeAheadComponent.inputTextControl.value).toBeFalsy();
  });

  it('should render vehicle list component when planet is selected', () => {
    fixture.detectChanges();
    const facadeService = fixture.componentRef.injector.get(
      FinderFacadeService
    );
    const injectedChangeDetector = fixture.componentRef.injector.get(
      ChangeDetectorRef as Type<ChangeDetectorRef>
    );

    facadeService.initializeAppData();
    component.planetSelected({ name: 'Donlon', distance: 100 });
    injectedChangeDetector.detectChanges();

    const vehicleListDebugElement = fixture.debugElement.query(
      By.directive(VehicleListComponent)
    );

    expect(vehicleListDebugElement).toBeTruthy();
    const vehicleListComponent = vehicleListDebugElement.componentInstance as VehicleListComponent;
    expect(vehicleListComponent.vehicleList === vehicleListToBeReturned);
    expect(vehicleListComponent.widgetId).toBe(component.widgetId);
    expect(vehicleListComponent.destinationDistance).toBe(
      component.destinationDistance
    );
    (expect(fixture) as any).toMatchSnapshot();
  });

  // it('should render vehicle list component when planet is selected', () => {
  //   // use this in beforeEach
  //   // fixture = TestBed.overrideComponent(DestinationWidgetComponent, {
  //   //   set: { host: { '(click)': 'dummy' } },
  //   // }).createComponent(DestinationWidgetComponent);

  //   fixture.detectChanges();
  //   const facadeService = fixture.componentRef.injector.get(
  //     FinderFacadeService
  //   );

  //   facadeService.initializeAppData();
  //   component.planetSelected({ name: 'Donlon', distance: 100 });

  //   // fake a click event so thta onpush component triggers change detection
  //   fixture.debugElement.triggerEventHandler('click', null);
  //   fixture.detectChanges();

  //   const vehicleListDebugElement = fixture.debugElement.query(
  //     By.directive(VehicleListComponent)
  //   );

  //   expect(vehicleListDebugElement).toBeTruthy();
  // });
});
