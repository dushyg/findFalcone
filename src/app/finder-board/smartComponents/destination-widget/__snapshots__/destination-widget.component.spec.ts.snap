// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DestinationWidgetComponent should render vehicle list component when planet is selected 1`] = `
<app-destination-widget
  changeDetector={[Function ViewRef_]}
  destinationDistance={[Function Number]}
  finderFacadeService={[Function FinderFacadeService]}
  isComponentActive={[Function Boolean]}
  lastSelectedPlanet={[Function String]}
  planetList={[Function Array]}
  planetListInitialized={[Function Boolean]}
  resetTypeAhead$={[Function Observable]}
  resetTypeAheadSubject={[Function Subject]}
  widgetId={[Function Number]}
>
  <div
    class="destinationWidget"
  >
    <span>
      Destination 3
    </span>
    <app-typeahead>
      <div
        class="typeahead-container"
      >
        <div
          class="typeadhead-wrapper"
        >
          <input
            class="typeAhead"
            placeholder="Type to search..."
            type="text"
          />
          
        </div>
      </div>
    </app-typeahead>
    
    <app-vehicle-list>
      <div
        class="vehicleList"
      >
        
        <div>
          <input
            id="vehicle_3_0"
            name="radio_3"
            type="radio"
            value="Space pod"
          />
            
          <label
            for="vehicle_3_0"
          >
             Space pod (2) 
          </label>
          <br />
        </div>
        <div>
          <input
            id="vehicle_3_1"
            name="radio_3"
            type="radio"
            value="Space rocket"
          />
            
          <label
            for="vehicle_3_1"
          >
             Space rocket (1) 
          </label>
          <br />
        </div>
        <div>
          <input
            id="vehicle_3_2"
            name="radio_3"
            type="radio"
            value="Space shuttle"
          />
            
          <label
            for="vehicle_3_2"
          >
             Space shuttle (1) 
          </label>
          <br />
        </div>
        <div>
          <input
            id="vehicle_3_3"
            name="radio_3"
            type="radio"
            value="Space ship"
          />
            
          <label
            for="vehicle_3_3"
          >
             Space ship (2) 
          </label>
          <br />
        </div>
      </div>
    </app-vehicle-list>
  </div>
</app-destination-widget>
`;
